src/support.js

```diff
@@ -246,14 +246,40 @@ Module['registerFunctions'] = registerFunctions;
 #endif // RELOCATABLE
 #endif // EMULATED_FUNCTION_POINTERS
 
+#if WASM_BACKEND
+#if RESERVED_FUNCTION_POINTERS
+var jsCallStartIndex = {{{ JSCALL_START_INDEX }}};
+var jsCallSigOrder = {{{ JSON.stringify(JSCALL_SIG_ORDER) }}};
+var jsCallNumSigs = Object.keys(jsCallSigOrder).length;
+var functionPointers = new Array(jsCallNumSigs * {{{ RESERVED_FUNCTION_POINTERS }}});
+#else
+var jsCallStartIndex = 1;
+var functionPointers = new Array({{{ RESERVED_FUNCTION_POINTERS }}});
+#endif
+#else
+var jsCallStartIndex = 1;
 var functionPointers = new Array({{{ RESERVED_FUNCTION_POINTERS }}});
+#endif
 
+#if WASM_BACKEND
+function addFunction(func, sig) {
+#else
 function addFunction(func) {
+#endif
 #if EMULATED_FUNCTION_POINTERS == 0
-  for (var i = 0; i < functionPointers.length; i++) {
+#if WASM_BACKEND
+#if RESERVED_FUNCTION_POINTERS
+  var base = jsCallSigOrder[sig] * {{{ RESERVED_FUNCTION_POINTERS }}};
+#else
+  var base = 0;
+#endif
+#else
+  var base = 0;
```

Comment:
Maybe a `//TODO: when '#if a || b' is supported, use '#if WASM_BACKEND || RESERVED_FUNCTION_POINTERS'`

This was really confusing to see at first glance.

Reply:
Done. Enclosed comments within `#if 0 ~ #endif` so that they wouldn't appear on the JS glue code.